pkg/builder/builder.go

```diff
@@ -0,0 +1,272 @@
+	loggerFactory                   logger.LoggerFactory
+	defaultProjectImage             string
+	defaultProjectUser              string
+	defaultProjectPostStartCommands []string
+}
+
+func NewBuilderFactory(config BuilderConfig) IBuilderFactory {
+	return &BuilderFactory{
+		daytonaServerConfigFolder:       config.DaytonaServerConfigFolder,
+		localContainerRegistryServer:    config.LocalContainerRegistryServer,
+		basePath:                        config.BasePath,
+		loggerFactory:                   config.LoggerFactory,
+		defaultProjectImage:             config.DefaultProjectImage,
+		defaultProjectUser:              config.DefaultProjectUser,
+		defaultProjectPostStartCommands: config.DefaultProjectPostStartCommands,
+	}
+}
+
+func (f *BuilderFactory) Create(p workspace.Project, cr *containerregistry.ContainerRegistry, gpc *gitprovider.GitProviderConfig) IBuilder {
+	uuid := uuid.New()
+	buildId := uuid.String()[:8]
+
+	builder := &Builder{
+		id:                buildId,
+		plugin:            nil,
+		project:           p,
+		containerRegistry: cr,
+		gitProviderConfig: gpc,
+
+		daytonaServerConfigFolder:       f.daytonaServerConfigFolder,
+		localContainerRegistryServer:    f.localContainerRegistryServer,
+		basePath:                        f.basePath,
+		loggerFactory:                   f.loggerFactory,
+		defaultProjectImage:             f.defaultProjectImage,
+		defaultProjectUser:              f.defaultProjectUser,
+		defaultProjectPostStartCommands: f.defaultProjectPostStartCommands,
+	}
+
+	return builder
+}
+
+func (b *Builder) GetBuilderPlugin() BuilderPlugin {
+	return b.plugin
+}
+
+func (b *Builder) Prepare() error {
+	hash, err := b.project.GetConfigHash()
+	if err != nil {
+		return err
+	}
+	b.hash = hash
+	projectDir := filepath.Join(b.basePath, hash, "project")
+
+	err = os.RemoveAll(projectDir)
+	if err != nil {
+		return err
+	}
+
+	gitservice := git.Service{
+		ProjectDir:        projectDir,
+		GitConfigFileName: "",
+		//	todo: write to project log
+		LogWriter: os.Stdout,
+	}
+
+	var auth *http.BasicAuth
+	if b.gitProviderConfig != nil {
+		auth = &http.BasicAuth{
+			Username: b.gitProviderConfig.Username,
+			Password: b.gitProviderConfig.Token,
+		}
+	}
+
+	err = gitservice.CloneRepository(&b.project, auth)
+	if err != nil {
+		return err
+	}
+
+	buildConfig := b.project.Build
+
+	if buildConfig != nil && *buildConfig == (workspace.ProjectBuild{}) {
+		//	detect is devcontainer
+		devcontainerPath := ".devcontainer/devcontainer.json"
+		isDevcontainer, err := fileExists(filepath.Join(projectDir, devcontainerPath))
+		if err != nil {
+			return err
+		}
+		if !isDevcontainer {
+			devcontainerPath = ".devcontainer.json"
+			isDevcontainer, err = fileExists(filepath.Join(projectDir, devcontainerPath))
+			if err != nil {
+				return err
+			}
+		}
+		if isDevcontainer {
+			buildConfig.Devcontainer = &workspace.ProjectBuildDevcontainer{
+				DevContainerFilePath: devcontainerPath,
+			}
+			goto initPlugin
```

Comment:
The `goto` seems pointless here because it just points to `if buildConfig.DevContainer != nil` which is again devcontainer specific. If `isDevcontainer` is false, that `initPlugin:` block will be executed again after we set the default project image.

To make this more maintainable, I suggest something like this:

```golang
         if buildConfig == nil || *buildConfig != (workspace.ProjectBuild{}) {
                return nil
         }

         // auto-detect config standard
         //	detect is devcontainer
	devcontainerPath := ".devcontainer/devcontainer.json"
	//       create a `isDevcontainer` function
	isDevcontainer, err := isDevcontainer(filepath.Join(projectDir, devcontainerPath))
	if err != nil {
		return err
	}
	if isDevcontainer {
		buildConfig.Devcontainer = &workspace.ProjectBuildDevcontainer{
			DevContainerFilePath: devcontainerPath,
		}
		b.plugin = &DevcontainerBuilder{
			DevcontainerBuilderConfig: DevcontainerBuilderConfig{
				buildId:                      b.id,
				project:                      b.project,
				loggerFactory:                b.loggerFactory,
				localContainerRegistryServer: b.localContainerRegistryServer,
				projectVolumePath:            filepath.Join(b.basePath, b.hash, "project"),
			},
		}
	        return nil
	}
	//	todo: detect dockerfile
	//	todo: detect nix

	//	no supported dev config standard found
	//	set default project image to ensure that project will run anyway
	b.project.Image = b.defaultProjectImage
	b.project.User = b.defaultProjectUser
	b.project.PostStartCommands = b.defaultProjectPostStartCommands

        return nil
```

Reply:
the goto statement feels discriminated :/

Reply:
You're not wrong 😄